src/java/org/apache/cassandra/index/sai/utils/RangeConcatIterator.java

```diff
@@ -78,50 +79,78 @@ public void close() throws IOException
     @SuppressWarnings("resource")
     protected PrimaryKey computeNext()
     {
-        while (!ranges.isEmpty())
+        if (currentRange == null || !currentRange.hasNext())
         {
-            RangeIterator current = ranges.poll();
-            if (current.hasNext())
-            {
-                PrimaryKey next = current.next();
-                // hasNext will update RangeIterator's current which is used to sort in PQ
-                if (current.hasNext())
-                    ranges.add(current);
-
-                return next;
-            }
+            if (!ranges.hasNext())
+                return endOfData();
+            currentRange = ranges.next();
+            assert currentRange.hasNext() : "Only non empty range iterators are added. Current range must not be empty.";
         }
-
-        return endOfData();
+        return currentRange.next();
     }
 
     public static Builder builder()
     {
-        return new Builder();
+        return builder(1);
+    }
+
+    public static Builder builder(int size)
+    {
+        return new Builder(size);
     }
 
     public static RangeIterator build(List<RangeIterator> tokens)
     {
-        return new Builder().add(tokens).build();
+        return new Builder(tokens.size()).add(tokens).build();
     }
 
     public static class Builder extends RangeIterator.Builder
     {
-        public Builder()
+        // We can use a list because the iterators are already in order
+        private final List<RangeIterator> rangeIterators;
+        public Builder(int size)
         {
             super(IteratorType.CONCAT);
+            this.rangeIterators = new ArrayList<>(size);
         }
 
-        protected RangeIterator buildIterator()
+        public int rangeCount()
         {
-            switch (rangeCount())
-            {
-                case 1:
-                    return ranges.poll();
+            return rangeIterators.size();
+        }
+
+        @Override
+        public Builder add(RangeIterator range)
```

Comment:
Could we add a check here that the range being added is in sorted order? I am concerned that leaving this as an unwritten expectation is not a good idea. It might also be worth adding something to the class javadoc explaining the sorting requirement of the ranges.

Reply:
Do you mean that the iterator itself is sorted or that the previous range's max is less than or equal to the next ranges min?

Reply:
The javadoc already comments on some aspect of sorting (though I find it somewhat ambiguous whether the RI's must be sorted or the list of RI's must be sorted):

https://github.com/datastax/cassandra/blob/06811f7c1cd6e364fa79fe98e17dffd55ba5bea2/src/java/org/apache/cassandra/index/sai/utils/RangeConcatIterator.java#L28-L29

FWIW, we already assert that the list itself is ordered here when we call `statistics.update(range);`:

https://github.com/datastax/cassandra/blob/49a51690caca180fc33bd49c7b95cfc8ee0627f3/src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java#L261-L281

We have a test that verifies the list is sorted here:

https://github.com/datastax/cassandra/blob/49a51690caca180fc33bd49c7b95cfc8ee0627f3/test/unit/org/apache/cassandra/index/sai/utils/RangeConcatIteratorTest.java#L32-L88